### DISCUSSION TITLE
Add documentation and update notes for Models, Records and Controllers

### DISCUSSION THREAD

**Discussion by lazlo-bonin** [craftcms/docs#299] [Ideas]
The Craft 2 plugin development documentation included information about Models, Records and Controllers which is no longer contained at all in the Craft 3 docs.

The current upgrade guide only mentions a very vague reference back to the Yii documentation for all of these concepts, without any specific instruction on how to port plugins that used them in Craft 2:

> See its comprehensive upgrade guide to learn about how things have changed under the hood.

Considering the big changes (e.g. using actual PHP attributes instead of `defineAttributes()` on models), this really hinders efforts to migrate plugins to Craft 3. It also led me to believe that Models and Records were no longer supported classes for Craft 3 plugins, until I delved into official P&T plugins like Digital Products to notice that they are indeed still used! This has us piecing up classes from the Craft 2 docs, Yii 2 upgrade docs, and reverse engineering... Far from ideal.

Here's the current table of core classes that are documented, which I'm making partly for myself to keep navigate through the upgrade process...

|Craft 2|Craft 3|Craft Update Notes|Yii Update Notes|
|-|-|-|-|
|[Models](https://docs.craftcms.com/v2/plugins/models.html)|**Missing**|**Missing**|[Models](https://www.yiiframework.com/doc/guide/2.0/en/intro-upgrade-from-v1#models)
|[Records](https://docs.craftcms.com/v2/plugins/records.html)|**Missing**|[DB Queries](https://docs.craftcms.com/v3/extend/updating-plugins.html#db-queries)|[Active Record](https://www.yiiframework.com/doc/guide/2.0/en/intro-upgrade-from-v1#active-record)
|[Controllers](https://docs.craftcms.com/v2/plugins/controllers.html)|**Missing**|(Multiple)|[Controllers](https://www.yiiframework.com/doc/guide/2.0/en/intro-upgrade-from-v1#controllers)
|[Element Actions](https://docs.craftcms.com/v2/plugins/element-actions.html)|[Element Actions](https://docs.craftcms.com/v3/extend/element-action-types.html)|Missing|N/A
|[Dashboard Widgets](https://docs.craftcms.com/v2/plugins/widgets.html)|[Widget Types](https://docs.craftcms.com/v3/extend/widget-types.html)|Missing|N/A
|[Field Types](https://docs.craftcms.com/v2/plugins/field-types.html)|[Field Types](https://docs.craftcms.com/v3/extend/field-types.html)|[Components](https://docs.craftcms.com/v3/extend/updating-plugins.html#components)|N/A
|[Services](https://docs.craftcms.com/v2/plugins/services.html)|[Services](https://docs.craftcms.com/v3/extend/services.html)|Missing|N/A
|[Template Variables](https://docs.craftcms.com/v2/plugins/variables.html)|[Extending Twig](https://docs.craftcms.com/v3/extend/extending-twig.html)|[Template Variables](https://docs.craftcms.com/v3/extend/updating-plugins.html#template-variables)|N/A

Is there an update planned for the core Models/Records/Controllers documentation, along with examples of how to create and use them in Craft 3?

**Comment by brandonkelly** [comment]
Yep we should have some mention of these things. However in Craft 3 models and records work the same as in Yii, which wasn’t the case in Craft 2, so the Yii documentation is the best place to start.
